Webhooks guide (#7663)
Rapptz/discord.py: docs/guide/topics/webhooks.rst

.. _guide_webhooks:

Webhooks
========

Webhooks are used for posting messages in Discord text channels without relying on a bot user.

The focus of this section is to provide a basic understanding of webhooks and
working with them using discord.py.

For detailed information about webhooks, See the documentation for :ref:`webhooks <discord-api-webhooks>`.

Introduction to webhooks
~~~~~~~~~~~~~~~~~~~~~~~~

The difference between sending messages with webhooks and bots is that bots require proper authentication and a bot user
whereas webhooks don't need either.

There are currently three types of webhooks:

1. Incoming webhooks
2. Channel follower webhooks
3. Application webhooks

**Incoming webhooks** are associated to a channel and can post messages in that channel
using a webhook token.

**Channel follower webhooks** are used for crossposting of messages in news channels
by Discord. This type does not have any token associated to it and bots cannot post
messages through these webhooks.

**Application webhooks** are used for creating interaction responses. The management of
this type of webhooks are handled internally by the library and is not in scope of this guide.

> This guide will primarily focus on the first type i.e Incoming webhooks.

Reviewer: ```suggestion
This guide will primarily focus on the first type, incoming webhooks.
```
"i.e" felt really out of place here.

Creating a webhook
~~~~~~~~~~~~~~~~~~

In order to create incoming webhooks in a channel, the member needs the
:attr:`~Permissions.manage_webhooks` permission in the given guild or text channel.

> To create a webhook, Navigate to "Server Settings > Integrations" page.

Reviewer: I feel whilst this is 100% correct, it should also mention that you can create webhooks on a channel too.
The "little cog" on the channel -> integrations -> create webhook

Author: Thanks, I have mentioned that. Let me know if there can be any improvement in the wording.

sure to copy it too.

The URL of webhook is used for executing the webhook and includes the webhook token
so don't store it in a public place.

You can also create webhooks through API using the :meth:`TextChannel.create_webhook` method.

Initializing webhooks
~~~~~~~~~~~~~~~~~~~~~

We'll now be initializing a webhook instance in our code.

There are two variants of webhooks provided by the library.

- :class:`Webhook`
- :class:`SyncWebhook`

The :class:`Webhook` is suitable for most cases as it is asynchronous while
:class:`SyncWebhook` is synchronous and is suitable for blocking environments. Both have the same
functionality with the difference of being sync and async. The HTTP methods usually return
the :class:`Webhook` class instance.

Webhooks are retrieved in two ways. They can be obtained through the :meth:`Guild.webhooks`
or :meth:`TextChannel.webhooks` methods. These methods requires authentication via a bot user.

The other method is to initialize "partial" webhooks using only webhook ID and token. This
method does not require authentication via bot user.

To create a partial webhook we would be using the URL that we copied earlier, passing it

> in the :meth:`~Webhook.from_url` method.

Reviewer: Do we intend to mention `Client.fetch_webhook` either here or in the section at L110?

Author: The latter seems a more suitable place for it. I have mentioned it there. Suggestions for wording would be appreciated. :>

.. currentmodule:: discord

.. _guide_webhooks:

Webhooks
========

Webhooks are used for posting messages in Discord text channels without relying on a bot user.

The focus of this section is to provide a basic understanding of webhooks and
working with them using discord.py.

For detailed information about webhooks, See the documentation for :ref:`webhooks <discord-api-webhooks>`.

Introduction to webhooks
~~~~~~~~~~~~~~~~~~~~~~~~

> The difference between sending messages with webhooks and bots is that latter requires
> proper authentication and a bot user while the former neither requires bot user nor
> authentication.

Reviewer 2: ```suggestion
The difference between sending messages with webhooks and bots is that bots require proper authentication and a bot user
whereas webhooks don't need either.
```

There are currently three types of webhooks:

1. Incoming webhooks
2. Channel follower webhooks
3. Application webhooks

**Incoming webhooks** are associated to a channel and can post messages in that channel
using a webhook token.

**Channel follower webhooks** are used for crossposting of messages in news channels
by Discord. This type does not have any token associated to it and bots cannot post
messages through these webhooks.

**Application webhooks** are used for creating interaction responses. The management of
this type of webhooks are handled internally by the library and is not in scope of this guide.

This guide will primarily focus on the first type i.e Incoming webhooks.

Creating a webhook
~~~~~~~~~~~~~~~~~~

In order to create incoming webhooks in a channel, the member needs the
:attr:`~Permissions.manage_webhooks` permission in the given guild or text channel.

> To create the a webhook, Navigate to "Server Settings > Integrations" page.

Reviewer 2: ```suggestion
To create a webhook, Navigate to "Server Settings > Integrations" page.
```

.. image:: /images/guide/webhooks/guild_settings_integrations.png
    :alt: Integrations Settings

Click "Create Webhook" to create a webhook.

.. image:: /images/guide/webhooks/webhook_created.png
    :alt: Webhook

> From here, you can edit the webhook's name and associated channel in which the webhook
> would be sending messages in. We'll be using the URL of webhook later in this guide so make
> sure to copy it too.

Reviewer 2: ```suggestion
From here, you can edit the webhook's name and the associated channel the webhook
will send messages in. We'll be using the URL of webhook later in this guide so make
sure to copy it too.
```

> The URL of webhook is used for executing the webhook and includes the webhook token
> so do not reveal this URL.

Reviewer 2: ```suggestion
The URL of webhook is used for executing the webhook and includes the webhook token
so don't store it in a public place.
```

You can also create webhooks through API using the :meth:`TextChannel.create_webhook` method.

Initializing webhooks
~~~~~~~~~~~~~~~~~~~~~

We'll now be initializing a webhook instance in our code.

There are two variants of webhooks provided by the library.

- :class:`Webhook`
- :class:`SyncWebhook`

> The :class:`Webhook` is suitable for most cases as it is an asynchronous counterpart while
> the other is synchronous and is suitable for blocking enivornments. Both have the same
> functionality with the difference of being sync and async. The HTTP methods usually return
> the :class:`Webhook` class instance.

Reviewer 2: ```suggestion
The :class:`Webhook` is suitable for most cases as it is asynchronous while
:class:`SyncWebhook` is synchronous and is suitable for blocking environments. Both have the same
functionality with the difference of being sync and async. The HTTP methods usually return
the :class:`Webhook` class instance.
```

Sending messages
~~~~~~~~~~~~~~~~

This is where the fun begins, Let's start posting messages through our webhook.
We use the :meth:`Webhook.send` method to send the messages.

.. code-block:: py

    async with aiohttp.ClientSession() as session:
        webhook = discord.Webhook.from_url("webhook-url", session=session)
        await webhook.send("Hello World")

This will produce the following output in Discord:

.. image:: /images/guide/webhooks/webhook_message.png
    :alt: Webhook message

Like normal messages, Webhooks support embeds, file attachments, allowed mentions and
other message features. An example of sending embeds is given below.

.. code-block:: py

    async with aiohttp.ClientSession() as session:
        webhook = discord.Webhook.from_url("webhook-url", session=session)
        embed = discord.Embed(title="Hello World", description="I'm a nice embed!")
        await webhook.send(embed=embed)

.. image:: /images/guide/webhooks/webhook_message_with_embed.png
    :alt: Webhook message with embed.

There are some features that are exclusive to webhook messages like hyperlinks in

> raw message's content.

Reviewer 3: This has changed relatively recently and is now possible within normal messages.

.. currentmodule:: discord

.. _guide_webhooks:

Webhooks
========

> Webhooks are used for posting messages in Discord text channels without relying on a bot user.

Reviewer 4: Err, the " or authentication" isn't strictly true, since whs still need tokens

The focus of this section is to provide a basic understanding of webhooks and
working with them using discord.py.

For detailed information about webhooks, see the documentation for :ref:`webhooks <discord-api-webhooks>`.

Introduction to Webhooks
~~~~~~~~~~~~~~~~~~~~~~~~

The difference between sending messages with webhooks and bots is that bots require proper
authentication and a bot user whereas webhooks don't need either.

There are currently three types of webhooks:

1. Incoming webhooks
2. Channel follower webhooks
3. Application webhooks

**Incoming webhooks** are associated to a channel and can post messages in that channel
using a webhook token.

**Channel follower webhooks** are used for crossposting of messages in news channels
by Discord. This type does not have any token associated to it and bots cannot post

> messages through these webhooks.

Reviewer 5: I don't really think that this example is much helpful and appropriate to be put here. I'll add a little bit more explanation for "News Channels" here.

Alternatively, webhooks can also be created from the "Integrations" page in a channel's settings.

.. image:: /images/guide/webhooks/guild_settings_integrations.png
    :alt: Integrations Settings

Click "Create Webhook" to create a webhook.

.. image:: /images/guide/webhooks/webhook_created.png
    :alt: Webhook

From here, you can edit the webhook's name and the associated channel the webhook
will send messages in. We'll be using the URL of webhook later in this guide so make
sure to copy it too.

The URL of webhook is used for executing the webhook and includes the webhook token
so don't store it in a public place.

You can also create webhooks through API using the :meth:`TextChannel.create_webhook` method.

Initializing Webhooks
~~~~~~~~~~~~~~~~~~~~~

We'll now be initializing a webhook instance in our code.

There are two variants of webhooks provided by the library.

- :class:`Webhook`
- :class:`SyncWebhook`

The :class:`Webhook` is suitable for most cases as it is asynchronous while
:class:`SyncWebhook` is synchronous and is suitable for blocking environments. Both have the same
functionality with the difference of being sync and async. The HTTP methods usually return

> the :class:`Webhook` class instance.

Reviewer 5: HTTP methods meant methods like `TextChannel.webhooks()` etc.
